aesd-char-driver/main.c

```diff
@@ -0,0 +1,261 @@
+    ssize_t read_offset = 0;
+
+    // Find the entry and offset for the given file position
+    read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(&(device->circle_buff), *f_pos, &read_offset);
+    if (!read_entry) {
+        goto error_handling;
+    }
+
+    // Check if bytes_to_read exceeds the available data
+    if (count > (read_entry->size - read_offset)) {
+            count = read_entry->size - read_offset;
+    }
+
+    // Read data using copy_to_user
+    ssize_t unread_count = copy_to_user(buf, (read_entry->buffptr + read_offset), count);
+
+    // Calculate the number of bytes successfully copied and update file_pos
+    retval = count - unread_count;
+    *f_pos += retval;
+
+error_handling:
+    mutex_unlock(&(device->lock));
+
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = -ENOMEM;
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+
+    struct aesd_dev *device;
+    const char *write_entry = NULL;
+    
+    
+    // Check for invalid parameters
+    if (!filp || !buf || !f_pos) {
+        // Invalid memory address
+        return -EFAULT;
+    }
+
+    if (count == 0) {
+        // Nothing to write
+        return 0;
+    }
+
+    
+    // Retrieve the AESD device from the file structure
+    device = (struct aesd_dev *)filp->private_data;
+
+    // lock mutex (interruptible)
+    if (mutex_lock_interruptible(&(device->lock))) {
+        PDEBUG(KERN_ERR "Failed to acquire the mutex lock");
+        return -ERESTARTSYS;
+    }
+
+    // memory allocation with kmalloc when no prior allocations present
+    if (0 == device->circle_buff_entry.size) {
+
+        device->circle_buff_entry.buffptr = kmalloc(count * sizeof(char), GFP_KERNEL);
+
+        if (!device->circle_buff_entry.buffptr) {
+            PDEBUG("malloc failed");
+            goto error_handling_write;
+        }
+    }
+    else {
+        // memory reallocation when prior allocations is present
+        device->circle_buff_entry.buffptr = krealloc(device->circle_buff_entry.buffptr, (device->circle_buff_entry.size + count) * sizeof(char), GFP_KERNEL);
+
+        if (!device->circle_buff_entry.buffptr) {
+            PDEBUG("realloc failed");
+            goto error_handling_write;
+        }
+    }
+
+    // Copy data from the user space buffer to the current command buffer
+    ssize_t  unwritten_count = copy_from_user((void *)(device->circle_buff_entry.buffptr + device->circle_buff_entry.size), buf, count);
+    retval = count - unwritten_count;
+    device->circle_buff_entry.size += retval;
+
+    // Check for '\n' character in the command; if found, add the entry to the circular buffer
+    if (memchr(device->circle_buff_entry.buffptr, '\n', device->circle_buff_entry.size)) {
+        
+        write_entry = aesd_circular_buffer_add_entry(&device->circle_buff, &device->circle_buff_entry);
+
+        if(write_entry) {
```

Comment:
@sane5805 it doesn't look like it's safe to always free when write_entry is not null, you need to modify `aesd_circular_buffer_add_entry` to return NULL anytime the buffer is not full